### PR TITLE
Weights in alpha for FocalLoss

### LINKED ISSUE #8601: Weights in alpha for FocalLoss
Currently, MONAI’s FocalLoss accepts alpha as a single scalar to balance background vs. foreground. This prevents specifying different weights for each class in multi-class segmentation, which is problematic for highly imbalanced datasets.
I would like FocalLoss to support alpha as a tensor or list of floats, allowing a separate weight for each class (e.g., [0.1, 0.3, 0.6] for a 3-class problem). Allowing per-class alpha directly would provide more precise control and better alignment with the original focal loss formulation.

### PR DESCRIPTION
**Pull request by ytl0623**
Fixes #8601

### Description

Support alpha as a list, tuple, or tensor of floats, in addition to the existing scalar support.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by ytl0623** [comment]
Hi @ericspod @KumoLiu @Nic-Ma ,

sorry to bother u. the CI is failing with `[Errno 28] No space left on device` during the build-docs and packaging steps. This seems to be a server-side issue. Could u plz re-trigger the workflows?

Thanks in advance!

File: monai/losses/focal_loss.py [hunk 1]
```diff
@@ -156,13 +164,19 @@ def forward(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
         loss: Optional[torch.Tensor] = None
         input = input.float()
         target = target.float()
+
+        alpha_arg: float | torch.Tensor | None = self.alpha
+        if isinstance(alpha_arg, torch.Tensor):
+            alpha_arg = alpha_arg.to(input.device)
```
Comment (ericspod) [inline monai/losses/focal_loss.py]: If this necessary here if `sigmoid_focal_loss` and `softmax_focal_loss` do this conversion as well?

**Comment by ericspod** [comment]
Hi @ytl0623 thanks for this, we should have the CICD issues sorted now. It looks good to me though I think we need to add tests for the class with a sequence `alpha` value to ensure this works correctly. Please add this then update the branch to incorporate the CICD fixes.

File: monai/losses/focal_loss.py (hunk 1 above)
Comment (ytl0623) [reply monai/losses/focal_loss.py]: You are right. This block is redundant, so I will remove it.

**Comment by ytl0623** [comment]
> Hi @ytl0623 thanks for this, we should have the CICD issues sorted now. It looks good to me though I think we need to add tests for the class with a sequence `alpha` value to ensure this works correctly. Please add this then update the branch to incorporate the CICD fixes.

Hi @ericspod, thanks for the feedback. I have removed the redundant conversion logic and added a new test case. It should be ready for another look. Thanks in advance!

File: tests/losses/test_focal_loss.py [hunk 2]
```diff
@@ -374,6 +374,47 @@ def test_script(self):
...
+    def test_alpha_sequence_broadcasting(self):
+        """
+        Test FocalLoss with alpha as a sequence for proper broadcasting.
+        """
+        num_classes = 3
+        alpha_seq = [0.1, 0.5, 2.0]
+        batch_size = 2
+        spatial_dims = (4, 4)
+
+        devices = ["cpu", "cuda"] if torch.cuda.is_available() else ["cpu"]
```
Comment (ericspod) [inline tests/losses/test_focal_loss.py]: We have [this definition](https://github.com/Project-MONAI/MONAI/blob/01711cf7e3af991aabc92cc3fdf4c0b65949e65e/tests/test_utils.py#L874C1-L874C14) to use in place of this line if you want. Another approach is to have a `device` argument and use the `parameterized.expand` decorator [like here](https://github.com/Project-MONAI/MONAI/blob/01711cf7e3af991aabc92cc3fdf4c0b65949e65e/tests/data/meta_tensor/test_meta_tensor.py#L201).

File: tests/losses/test_focal_loss.py [hunk 3]
```diff
@@ -374,6 +374,47 @@ def test_script(self):
             test_input = torch.ones(2, 2, 8, 8)
             test_script_save(loss, test_input, test_input)
 
+    def test_alpha_sequence_broadcasting(self):
```
Comment (ericspod) [inline tests/losses/test_focal_loss.py]: Here I think we do need another test, or use `parameterized.expand`, to test with and without the background.

**Comment by ericspod** [comment]
Just about there, let's add a test for background=False and the stacklevel change for the warning and we'll be good to go.

File: tests/losses/test_focal_loss.py (hunk 3 above)
Comment (ytl0623) [reply tests/losses/test_focal_loss.py]: Done.

**Comment by ytl0623** [comment]
Thank you for the feedback. I have refactored the tests to use `@parameterized.expand`.

**Review by ericspod — APPROVED** [review]
Looks good now!

**Review by ericspod — APPROVED** [review]
